### PR TITLE
Goa 2457 independent annotation indexing jobs

### PR DESCRIPTION
**Pull request by TonyWardell**
Allow Annotation Indexing + Coterms creation OR just Coterms creation.
Controlled by the spring setting the spring.batch.job.names in the index. variables properties file within the indexing deployments directory.

### CODE REVIEW

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/OutputPath.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.google.common.base.Preconditions;
+import java.util.Objects;
+import org.springframework.core.io.FileSystemResource;
+
+/**
+ * @author Tony Wardell
+ * Date: 06/02/2017
+ * Time: 16:31
+ * Created with IntelliJ IDEA.
+ */
```
Comment (eddturner) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/OutputPath.java]: The class name is a bit misleading (I'd expected it to be something to do with a path, but it's a resource wrapper). I think it'd help if it was named something a bit clearer and have accompanying javadoc to help a future developer use it.

File: deployments/src/main/distros/solr/index/bin/annotation/index.variables [hunk 2]
```diff
@@ -4,8 +4,8 @@
 
 ## ---------------------------- ASSOCIATED SOLR CORE INFO ----------------------------
 core_name="annotation";
-job_name="annotationIndexingJob";
-
+job_name="annotationIndexingJob"; #index annotations and coterms
+#job_name="coTermIndexingJob"; to run coterms only
```
Comment (eddturner) [inline deployments/src/main/distros/solr/index/bin/annotation/index.variables]: has this been tested on dev?

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/OutputPath.java (hunk 1 above)
Comment (TonyWardell) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/OutputPath.java]: Removed this class.

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/OutputPath.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+public class OutputPath {
```
Comment (TonyWardell) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/OutputPath.java]: Removed this class.

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/CoTermsConfig.java [hunk 4]
```diff
@@ -48,11 +59,53 @@
...
 
+    @Value("${indexing.coterms.chunk.size:1}")
+    private int cotermsChunk;
+    @Value("${indexing.coterm.loginterval:1000}")
+    private int coTermLogInterval;
+    @Value("${indexing.coterms.manual:#{systemProperties['user.dir']}/QuickGO/CoTermsManual}")
+    String manualCoTermsPath;
+    @Value("${indexing.coterms.all:#{systemProperties['user.dir']}/QuickGO/CoTermsAll}")
+    String allCoTermsPath;
+
```
Comment (TonyWardell) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/CoTermsConfig.java]: Made the logger static.

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/CoTermsConfig.java [hunk 5]
```diff
@@ -127,20 +176,18 @@ public CoTermsProcessor coTermsAllCalculator(
...
+    private ItemWriteListener<QuickGODocument> logWriteRateListener(final int writeInterval) {
+        return new ItemRateWriterListener<>(Instant.now(), writeInterval);
+    }
+
+    private StepExecutionListener logStepListener() {
+        return new LogStepListener();
+    }
 
-        }
+    //todo is this signature correct?
```
Comment (TonyWardell) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/coterms/CoTermsConfig.java]: Fixed SkipLoggerListener generic.

File: indexing/src/test/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingBatchIT.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import static uk.ac.ebi.quickgo.index.annotation.coterms.CoTermsConfig.CO_TERM_ALL_SUMMARIZATION_STEP;
+import static uk.ac.ebi.quickgo.index.annotation.coterms.CoTermsConfig.CO_TERM_MANUAL_SUMMARIZATION_STEP;
+
+/**
+ * Tests whether Spring Batch is correctly wired up to run the annotation indexing.
+ *
+ * Created 22/04/16
+ * @author Edd
+ */
+@ActiveProfiles(profiles = {"embeddedServer"})
```
Comment (TonyWardell) [inline indexing/src/test/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingBatchIT.java]: Done.

File: indexing/src/test/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingBatchIT.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Before
+    public void setUp() {
+        annotationRepository.deleteAll();
+    }
+
+
+    @Test
+    public void successfulCoTermsOnlyJob() throws Exception {
+        JobExecution jobExecution = jobLauncherTestUtils.launchJob();
```
Comment (TonyWardell) [inline indexing/src/test/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingBatchIT.java]: Removed annotation repository instances.

File: deployments/src/main/distros/solr/index/bin/annotation/index.variables (hunk 2 above)
Comment (TonyWardell) [inline deployments/src/main/distros/solr/index/bin/annotation/index.variables]: No.

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingConfig.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import org.springframework.core.io.Resource;
+import org.springframework.data.solr.core.SolrTemplate;
+
+/**
+ * Sets up batch jobs for annotation indexing.
+ *
+ * Created 20/04/16
+ * @author Edd
+ */
+@Configuration
```
Comment (TonyWardell) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingConfig.java]: I'd already done that for this class - what's the issue here?

File: deployments/src/main/distros/solr/index/bin/annotation/index.variables (hunk 2 above)
Comment (eddturner) [inline deployments/src/main/distros/solr/index/bin/annotation/index.variables]: would be great to test before merging

File: indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingConfig.java (hunk 8 above)
Comment (eddturner) [inline indexing/src/main/java/uk/ac/ebi/quickgo/index/annotation/CoTermIndexingConfig.java]: When I reviewed this yesterday, it wasn't formatted:

```java
... end of class
    return new SkipLoggerListener<>();
    }
    }
```
But it was reformatted 11 mins ago, thanks.

**Comment by TonyWardell** [comment]
Changes made, ran coterms alone indexing and coterms + annotation indexing against dev.
Files output and indexing OK

**Pull request merged**